src/sources/util/unix.rs

```diff
@@ -0,0 +1,71 @@
+use crate::event::Event;
+use crate::sources::Source;
+use bytes::Bytes;
+use futures::{future, sync::mpsc, Future, Sink, Stream};
+use std::{fs::canonicalize, path::PathBuf};
+use tokio::{
+    self,
+    codec::{FramedRead, LinesCodec},
+};
+use tokio_uds::UnixListener;
+use tracing::field;
+use tracing_futures::Instrument;
+
+/**
+* A function that takes a single line of a received message and builds an Event object.
+* Used by build_unix_source to encode any logic necessary for transforming the raw message
+* into a usuable Event object (for example decoding syslog messages).
+**/
+type BuildEventFn = fn(&str, Option<Bytes>, &str) -> Option<Event>;
+
+/**
+* Returns a Source object corresponding to a Unix domain socket.  Passing in different functions
+* for build_event can allow for different source-specific logic (such as decoding syslog messages
+* in the syslog source).
+**/
+pub fn build_unix_source<'a>(
+    path: PathBuf,
+    max_length: usize,
+    host_key: String,
+    out: mpsc::Sender<Event>,
+    build_event: BuildEventFn,
+) -> Source {
+    let out = out.sink_map_err(|e| error!("error sending line: {:?}", e));
+
+    Box::new(future::lazy(move || {
+        let listener = UnixListener::bind(&path).expect("failed to bind to listener socket");
+
+        let path = canonicalize(path).expect("could not canonicalize path");
```

Comment:
On the other hand, if there is an error regarding the path, the path in any error messages will be different from what is in the configuration file, which may cause confusion for users, so I'm doubtful as to it's value.

In either case, should the path not be canonicalized *before* you listen to it? If so, it should go outside of the future closure in order to catch any configuration errors when the source configuration is being handled rather than when trying to connect.

Reply:
Ah yeah good call, I just went ahead and removed it.